<a href="https://colab.research.google.com/github/alessandrossC/Detecting_Fake_News/blob/main/02_baseline.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import os
import pandas as pd
import numpy as np
import joblib
import xgboost as xgb
from sklearn.dummy import DummyClassifier
from sklearn.metrics import log_loss, classification_report, accuracy_score
from scipy.sparse import hstack, csr_matrix
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
DATA_PATH = '/content/processed'
train_df = pd.read_csv(os.path.join(DATA_PATH, 'train.csv.zip'), compression='zip')
val_df = pd.read_csv(os.path.join(DATA_PATH, 'val.csv.zip'), compression='zip')

In [ ]:
model = DummyClassifier(strategy='most_frequent', random_state=7)
model.fit(train_df[['title']], train_df['is_fake'])

preds_proba = model.predict_proba(val_df[['title']])
preds = model.predict(val_df[['title']])

print('Log loss:', log_loss(val_df['is_fake'], preds_proba))
print(classification_report(val_df['is_fake'], preds, zero_division=0))

Log loss: 17.111786208233823
              precision    recall  f1-score   support

           0       0.00      0.00      0.00      3394
           1       0.53      1.00      0.69      3755

    accuracy                           0.53      7149
   macro avg       0.26      0.50      0.34      7149
weighted avg       0.28      0.53      0.36      7149



This baseline sets the minimum performance threshold. Any meaningful machine learning model should significantly outperform this result, demonstrating its ability to distinguish between fake and real news.

In [ ]:
model = DummyClassifier(strategy='uniform', random_state=7)
model.fit(train_df[['title']], train_df['is_fake'])

preds_proba = model.predict_proba(val_df[['title']])
preds = model.predict(val_df[['title']])

print('Log loss:', log_loss(val_df['is_fake'], preds_proba))
print(classification_report(val_df['is_fake'], preds, zero_division=0))

Log loss: 0.6931471805599454
              precision    recall  f1-score   support

           0       0.47      0.50      0.49      3394
           1       0.52      0.50      0.51      3755

    accuracy                           0.50      7149
   macro avg       0.50      0.50      0.50      7149
weighted avg       0.50      0.50      0.50      7149



The DummyClassifier with the 'uniform' strategy predicts each class at random, regardless of the input data. As expected, the accuracy, precision, recall, and F1-score for both classes are around 0.50. The log loss is approximately 0.69, which matches the theoretical value for random guessing in a binary classification problem. This result sets a reference point for model performance; any meaningful model should achieve significantly better metrics.

In [ ]:
feature_info = joblib.load(os.path.join(DATA_PATH, 'feature_info.joblib'))

In [ ]:
numerical_features = feature_info['numerical_features']
text_features = feature_info['text_features']

In [ ]:
train_target = train_df['is_fake']
val_target = val_df['is_fake']

In [ ]:
def prepare_vectorizer(
        df,
        text_col1='title_clean',
        text_col2='text_clean',
        **vectorizer_params):
    """
    Fit a TfidfVectorizer on two text columns together.
    """
    default_params = {'max_features': 10000, 'ngram_range': (1, 2)}
    default_params.update(vectorizer_params)
    all_text = pd.concat([df[text_col1], df[text_col2]]).fillna('')
    vectorizer = TfidfVectorizer(**default_params)
    vectorizer.fit(all_text)
    return vectorizer

def transform(df, text_col1, text_col2, vectorizer):
    """
    Transform new data using fitted vectorizer.
    """
    vec1 = vectorizer.transform(df[text_col1].fillna(''))
    vec2 = vectorizer.transform(df[text_col2].fillna(''))
    return vec1, vec2

def combine_features(df, numerical_features, vec1, vec2):
    """
    Combine numerical features with vectorized text features.
    """
    return hstack([csr_matrix(df[numerical_features].values), vec1, vec2])

In [ ]:
vectorizer = prepare_vectorizer(train_df, text_col1='title_clean', text_col2='text_clean')

In [ ]:
train_title_vec, train_text_vec = transform(train_df, 'title_clean', 'text_clean', vectorizer)
val_title_vec, val_text_vec = transform(val_df, 'title_clean', 'text_clean', vectorizer)

In [ ]:
numerical_features = feature_info['numerical_features']
X_train = combine_features(train_df, numerical_features, train_title_vec, train_text_vec)
X_val = combine_features(val_df, numerical_features, val_title_vec, val_text_vec)

In [ ]:
y_train = train_df['is_fake']
y_val = val_df['is_fake']

In [ ]:
model = LogisticRegression(max_iter=1000, class_weight='balanced', random_state=42)
model.fit(X_train, y_train)
y_pred = model.predict(X_val)
print("Accuracy:", accuracy_score(y_val, y_pred))
print(classification_report(y_val, y_pred, digits=3))

Accuracy: 0.9925863757168835
              precision    recall  f1-score   support

           0      0.991     0.994     0.992      3394
           1      0.994     0.991     0.993      3755

    accuracy                          0.993      7149
   macro avg      0.992     0.993     0.993      7149
weighted avg      0.993     0.993     0.993      7149



/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [17]:
#TF-IDF Vectorizer
vectorizer = prepare_vectorizer(
    train_df,
    text_col1='title_clean',
    text_col2='text_clean',
    max_features=10000
)

train_title_vec, train_text_vec = transform(
    train_df,
    text_col1='title_clean',
    text_col2='text_clean',
    vectorizer=vectorizer
)

numerical_features = feature_info['numerical_features']

X_train = combine_features(
    train_df,
    numerical_features=numerical_features,
    vec1=train_title_vec,
    vec2=train_text_vec
)


In [18]:
val_title_vec, val_text_vec = transform(
    val_df,
    text_col1='title_clean',
    text_col2='text_clean',
    vectorizer=vectorizer
)

X_val = combine_features(
    val_df,
    numerical_features=numerical_features,
    vec1=val_title_vec,
    vec2=val_text_vec
)

In [19]:
from sklearn.metrics import log_loss

def evaluate_model(X_train, X_val, y_train, y_val, model):
    y_pred_train = model.predict_proba(X_train)
    y_pred_val = model.predict_proba(X_val)
    loss_train = log_loss(y_train, y_pred_train)
    loss_val = log_loss(y_val, y_pred_val)
    return loss_train, loss_val


In [20]:
train_title_vec, train_text_vec = transform(
    train_df,
    text_col1='title_clean',
    text_col2='text_clean',
    vectorizer=vectorizer
)

train_inputs = combine_features(
    train_df,
    numerical_features=numerical_features,
    vec1=train_title_vec,
    vec2=train_text_vec
)

In [21]:
val_title_vec, val_text_vec = transform(
    val_df,
    text_col1='title_clean',
    text_col2='text_clean',
    vectorizer=vectorizer
)

val_inputs = combine_features(
    val_df,
    numerical_features=numerical_features,
    vec1=val_title_vec,
    vec2=val_text_vec
)


In [22]:
model_lr = LogisticRegression(
    random_state=7,
    max_iter=1000,
    class_weight='balanced',
    n_jobs=-1
).fit(train_inputs, train_target)

log_loss_train, log_loss_val = evaluate_model(
    train_inputs, val_inputs, train_target, val_target, model_lr
)
print('Log loss on Train set:', log_loss_train)
print('Log loss on Validation set:', log_loss_val)

Log loss on Train set: 0.02731060745113039
Log loss on Validation set: 0.031005534227412444


In [23]:
y_val_pred = model_lr.predict(val_inputs)
print("Accuracy:", accuracy_score(val_target, y_val_pred))
print(classification_report(val_target, y_val_pred, digits=3))

Accuracy: 0.9921667366065184
              precision    recall  f1-score   support

           0      0.990     0.993     0.992      3394
           1      0.994     0.991     0.993      3755

    accuracy                          0.992      7149
   macro avg      0.992     0.992     0.992      7149
weighted avg      0.992     0.992     0.992      7149



The Logistic Regression model, using a combination of TF-IDF text features and numerical engineered features, achieves extremely high performance on the validation set:

Accuracy: 99.2%

Precision, Recall, and F1-score: all around 0.99 for both real and fake news classes.

This demonstrates the model's exceptional ability to distinguish between fake and real news in this dataset. The results significantly outperform all baselines, confirming that the chosen features and preprocessing steps are highly effective.

In [24]:
experiment_results = []
results = {}
results['model'] = 'Logistic Regression'
results['log_loss_train'] = log_loss_train
results['log_loss_val'] = log_loss_val

experiment_results.append(results)

In [25]:
model_rf = RandomForestClassifier(
    n_estimators=10,
    class_weight='balanced',
    n_jobs=-1,
    random_state=42
).fit(train_inputs, train_target)

log_loss_train, log_loss_val = evaluate_model(
    train_inputs, val_inputs, train_target, val_target, model_rf
)
print('Log loss on Train set:', log_loss_train)
print('Log loss on Validation set:', log_loss_val)

Log loss on Train set: 0.030701452032678118
Log loss on Validation set: 0.08767645515555846


The RandomForestClassifier achieves a log loss of ~0.03 on the training set and ~0.09 on the validation set.
This performance is slightly worse than Logistic Regression on the validation set, indicating that for this specific combination of engineered and text features, Logistic Regression remains the stronger baseline. Further improvements may be possible with boosting algorithms such as XGBoost or LightGBM, or by exploring more advanced neural architectures.

In [26]:
model_rf = RandomForestClassifier(
    n_estimators=10,
    max_depth=10,
    class_weight='balanced',
    n_jobs=-1
).fit(train_inputs, train_target)

log_loss_train, log_loss_val = evaluate_model(
    train_inputs, val_inputs, train_target, val_target, model_rf
)
print('Log loss on Train set:', log_loss_train)
print('Log loss on Validation set:', log_loss_val)

Log loss on Train set: 0.2268905329434506
Log loss on Validation set: 0.23625998777838195


In [27]:
model_rf = RandomForestClassifier(
    n_estimators=10,
    max_depth=20,
    class_weight='balanced',
    n_jobs=-1
).fit(train_inputs, train_target)

log_loss_train, log_loss_val = evaluate_model(
    train_inputs, val_inputs, train_target, val_target, model_rf
)
print('Log loss on Train set:', log_loss_train)
print('Log loss on Validation set:', log_loss_val)

Log loss on Train set: 0.12272193558554284
Log loss on Validation set: 0.1551066919834665


In [28]:
model_rf = RandomForestClassifier(
    n_estimators=10,
    max_leaf_nodes=1000,
    class_weight='balanced',
    n_jobs=-1
).fit(train_inputs, train_target)

log_loss_train, log_loss_val = evaluate_model(
    train_inputs, val_inputs, train_target, val_target, model_rf
)
print('Log loss on Train set:', log_loss_train)
print('Log loss on Validation set:', log_loss_val)

Log loss on Train set: 0.039345427124300385
Log loss on Validation set: 0.0828805959466344


In [29]:
model_rf = RandomForestClassifier(
    n_estimators=10,
    max_leaf_nodes=1500,
    class_weight='balanced',
    n_jobs=-1
).fit(train_inputs, train_target)

log_loss_train, log_loss_val = evaluate_model(
    train_inputs, val_inputs, train_target, val_target, model_rf
)
print('Log loss on Train set:', log_loss_train)
print('Log loss on Validation set:', log_loss_val)

Log loss on Train set: 0.02953446240340558
Log loss on Validation set: 0.08248911134448124


In [30]:
model_rf = RandomForestClassifier(
    n_estimators=10,
    max_leaf_nodes=1750,
    class_weight='balanced',
    n_jobs=-1
).fit(train_inputs, train_target)

log_loss_train, log_loss_val = evaluate_model(
    train_inputs, val_inputs, train_target, val_target, model_rf
)
print('Log loss on Train set:', log_loss_train)
print('Log loss on Validation set:', log_loss_val)

Log loss on Train set: 0.032718098483617356
Log loss on Validation set: 0.09344234582393324


In [31]:
results = {}
results['model'] = 'Random Forest'
results['log_loss_train'] = log_loss_train
results['log_loss_val'] = log_loss_val

experiment_results.append(results)

In [32]:
pd.DataFrame(experiment_results)

model  log_loss_train  log_loss_val
0  Logistic Regression        0.027311      0.031006
1        Random Forest        0.032718      0.093442

Logistic Regression should be used as the main baseline for further experiments. Future work could focus on boosting algorithms (like XGBoost or LightGBM) or neural architectures to see if they can further improve upon this strong baseline.

In [33]:
joblib.dump(experiment_results, os.path.join(DATA_PATH, 'experiment_results.joblib'))

['/content/processed/experiment_results.joblib']

In [34]:
#XGBoost
scale_pos_weight = np.round(sum(train_target == 0) / sum(train_target == 1), 4)
scale_pos_weight

np.float64(0.9038)

In [35]:
model_xgb = XGBClassifier(
    n_estimators = 10,
    learning_rate = 0.1,
    random_state = 7,
    n_jobs = -1,
    eval_metric = 'logloss',
    tree_method = 'hist',
    scale_pos_weight = scale_pos_weight,
).fit(train_inputs, train_target)

log_loss_train, log_loss_val = evaluate_model(
    train_inputs, val_inputs, train_target, val_target, model_xgb
)
print('Log loss on Train set:', log_loss_train)
print('Log loss on Validation set:', log_loss_val)

Log loss on Train set: 0.1997330535537069
Log loss on Validation set: 0.20010054250054632


In [36]:
model_xgb = XGBClassifier(
    n_estimators = 10,
    max_depth = 10,
    learning_rate = 0.1,
    random_state = 7,
    n_jobs = -1,
    eval_metric = 'logloss',
    tree_method = 'hist',
    scale_pos_weight = scale_pos_weight,
).fit(train_inputs, train_target)

log_loss_train, log_loss_val = evaluate_model(
    train_inputs, val_inputs, train_target, val_target, model_xgb
)
print('Log loss on Train set:', log_loss_train)
print('Log loss on Validation set:', log_loss_val)

Log loss on Train set: 0.1993947482807029
Log loss on Validation set: 0.2000494639814533


The best result for Logistic Regression.

In [37]:
def get_match_words(q1, q2):
    """
    Find common words between two text fields (e.g., title and text).
    Returns:
        match_words: string of common words (space-separated)
        count_match_words: number of common words
    """
    try:
        q1 = set(str(q1).split())
        q2 = set(str(q2).split())
    except Exception as e:
        q1 = set()
        q2 = set()
    match_words_list = list(q1 & q2)
    count_match_words = len(match_words_list)
    match_words = ' '.join(match_words_list) if count_match_words > 0 else ''
    return match_words, count_match_words

In [38]:
def train_evaluate_model(
    df_train, df_val,
    numerical_features, text_col1, text_col2,
    vectorizer, model, y_train, y_val
):
    """
    Тренує та оцінює модель на числових і текстових фічах одночасно.
    """
    # Fit vectorizer на train
    vectorizer.fit(pd.concat([df_train[text_col1], df_train[text_col2]]).fillna(''))

    # Трансформуй текст
    train_text1_vec = vectorizer.transform(df_train[text_col1].fillna(''))
    train_text2_vec = vectorizer.transform(df_train[text_col2].fillna(''))
    val_text1_vec = vectorizer.transform(df_val[text_col1].fillna(''))
    val_text2_vec = vectorizer.transform(df_val[text_col2].fillna(''))

    # Комбінуй із числовими ознаками
    X_train = hstack([csr_matrix(df_train[numerical_features].values), train_text1_vec, train_text2_vec])
    X_val = hstack([csr_matrix(df_val[numerical_features].values), val_text1_vec, val_text2_vec])

    # Train
    model.fit(X_train, y_train)

    # Log loss
    log_loss_train, log_loss_val = evaluate_model(X_train, X_val, y_train, y_val, model)

    return log_loss_train, log_loss_val


In [39]:
X_train = pd.DataFrame()
X_val = pd.DataFrame()

X_train['match_words'] = train_df.apply(
    lambda x: get_match_words(x['title_clean'], x['text_clean'])[0], axis=1)
X_val['match_words']  = val_df.apply(
    lambda x: get_match_words(x['title_clean'], x['text_clean'])[0], axis=1)

y_train = train_df['is_fake']
y_val = val_df['is_fake']

In [42]:
vectorizer = TfidfVectorizer(
    stop_words='english',
    lowercase=True,
    analyzer='word',
    max_features=10000
)
model = LogisticRegression(
    max_iter=1000,
    random_state=7,
    class_weight='balanced',
    n_jobs=-1
)

log_loss_train, log_loss_val = train_evaluate_model(
    train_df, val_df,
    numerical_features, 'title_clean', 'text_clean',
    vectorizer, model, y_train, y_val
)

print('Log loss on Train set:', log_loss_train)
print('Log loss on Validation set:', log_loss_val)

Log loss on Train set: 0.02765740580529381
Log loss on Validation set: 0.03104545215264795


Using only the "match_words" feature (the intersection of words between the title and text), the model achieved a log loss of ~0.03 on both the training and validation sets.
This result matches the performance of the full feature set, suggesting that this single feature is highly predictive in the current dataset.

In [43]:
experiment_results = joblib.load(os.path.join(DATA_PATH, 'experiment_results.joblib'))
results = {}
results['model'] = 'Logistic Regressin with TF-IDF on matching words'
results['log_loss_train'] = np.round(log_loss_train, 5)
results['log_loss_val'] = np.round(log_loss_val, 5)

experiment_results.append(results)

joblib.dump(experiment_results, os.path.join(DATA_PATH, 'experiment_results.joblib'))

['/content/processed/experiment_results.joblib']